src/Hooks.ts

```diff
@@ -17,18 +17,69 @@ import {
   EventType,
 } from 'react-navigation';
 
-export function useNavigation<S>(): NavigationScreenProp<S & NavigationRoute> {
+function useNavigationSafe<S>(): NavigationScreenProp<S & NavigationRoute> {
   const navigation = useContext(NavigationContext) as any; // TODO typing?
+
   if (!navigation) {
     throw new Error(
       "react-navigation hooks require a navigation context but it couldn't be found. " +
         "Make sure you didn't forget to create and render the react-navigation app container. " +
         'If you need to access an optional navigation object, you can useContext(NavigationContext), which may return'
     );
   }
+
   return navigation;
 }
 
+function useStableCallback(cb: (...args: any) => any) {
+  const ref = useRef(cb);
+
+  useLayoutEffect(() => {
+    ref.current = cb;
+  }, [cb]);
```

Comment:
Hm, I might be missing something, but do we need to run this as an effect rather than just straight-up `ref.current = cb;`? The `ref.current` in the function passed to `useCallback` should refer to the value-at-time-of-invocation anyway, so I don't think it should lead to staleness.

Reply:
Good point, by default I put all the side effects on the commit phase or layout phase. But maybe in this case is not necessary and it can be runned on the render phase as it's only a reasigment and don't interfers on the render.

Maybe changing the ref on render phase can be considered an unexpected side effect?
https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects

Reply:
oh and this pattern is already used on the `useGetter` function below.

Reply:
Hi,

Doing this kind of side effect looks unsafe to me with concurrent mode. React might decide to start rendering something based on predictions, to warm up the next view, and that transition may end up to never happen. Dan Abramov mentioned that recently on Twitter.

I think it's safer and good enough to use an effect for this usecase